File: 2022/day01/solution.py
"""
Day 01: Calorie Counting

Part 1: Get the amount of calories the elf who is carrying the most has.
Part 2: Get the top 3 elves who hold the most calories, and sum them.
"""

def parse_data(f):
    return sorted([sum(list(map(int,a.split('\n')))) for a in f.read().split('\n\n')], reverse=True)

def part1(data):
    return data[0]

def part2(data):
    return sum(data[:3])

File: 2022/day01/solution.rs
use std::fs;

fn parse_data(data: &String) -> Vec<u32> {
    let split = data.split("\n\n");

    let mut elves = Vec::new();
    for elf in split {
        let cals = elf.lines().map(|x| x.parse::<u32>().unwrap()).sum();
        
        elves.push(cals);
    }

    elves.sort();

    return elves;
}

fn part1(data: &Vec<u32>) -> u32 {
    return *data.get(data.len()-1).unwrap();
}

fn part2(data: &Vec<u32>) -> u32 {
    return data.get(data.len()-4..data.len()-1).unwrap().iter().sum::<u32>();
}

fn main() {
    let dummy: String = fs::read_to_string("data/2022/day01/dummy.txt")
        .expect("Shouldn't happen");
    let dummy_data = parse_data(&dummy);
    let real: String = fs::read_to_string("data/2022/day01/data.txt")
        .expect("Shouldn't happen");
    let real_data = parse_data(&real);

    println!("Part One:\nDummy Data: {}\nReal Data: {}\n", part1(&dummy_data), part1(&real_data));

    println!("Part Two:\nDummy Data: {}\nReal Data: {}", part2(&dummy_data), part2(&real_data));
}